Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -151,6 +151,7 @@ extension SOCKSClientHandler {
         // If we have any buffered writes then now
         // we can send them.
         self.writeBufferedData(context: context)
+        context.fireUserInboundEventTriggered(SOCKSProxyEstablishedEvent())
```

Comment:
Can I propose that we fire this _first_? Otherwise we will be delivering application bytes before we've fired this user event, which makes managing things very hard.

Reply:
If we fire this event first, we might remove the ChannelHandler before all data is written... I think with this, we def. need to add the wanted removal to the state machine.

Reply:
It's ok to remove the channel handler first. There are two important notes to channel handler removal:

1. No-one can force a handler out of the channel before it's willing to go
2. During `handlerRemoved` you still have a context that can be used to forward data.